Do not reload code inside gr.NO_RELOAD context

freddyaboulton opened the pull request:
## Description

Adds a `gr.NO_RELOAD` symbol. Any code in a `if gr.NO_RELOAD` code-block will not be re-evaluated when the source file is reloaded. Helpful for importing modules that do not like to be reloaded (`tiktoken`, `numpy`) as well as database connections and long running set up code.

Sample usage:

```python
import gradio as gr
import os

if gr.NO_RELOAD:
    print("IMPORTING")
    import tiktoken
    import numpy as np

def image_mod(image):
    print(np.zeros(5))
    return image.rotate(45)


cheetah = os.path.join(os.path.dirname(__file__), "images/lion.jpg")

demo = gr.Interface(image_mod, gr.Image(type="pil", value=cheetah, label="Input"), "image",
    flagging_options=["blurry", "incorrect", "other"], title="Testing a title!", examples=[
        os.path.join(os.path.dirname(__file__), "images/lion.jpg"),
        os.path.join(os.path.dirname(__file__), "images/cheetah1.jpg")
        ])

if __name__ == "__main__":
    demo.launch()
```

Attached gif below, note that the "IMPORTING" statement is not printed during reloads (only when first running the file as expected). 

![reload_no_import](https://github.com/gradio-app/gradio/assets/41651716/eb9c6fc1-182e-42d2-bb7b-520a66a8539e)

Also tested:
- modifying a file imported in the main demo file. Works as expected.
- `gradio cc dev`. Works as expected
- Also adds a reload mode test


Closes: #5402 
Closes: #7123 
Closes #6214

## 🎯 PRs Should Target Issues

Before your create a PR, please check to see if there is [an existing issue](https://github.com/gradio-app/gradio/issues) for this change. If not, please create an issue before you create this PR, unless the fix is very small. 

Not adhering to this guideline will result in the PR being closed. 

## Tests

1. PRs will only be merged if tests pass on CI. To run the tests locally, please set up [your Gradio environment locally](https://github.com/gradio-app/gradio/blob/main/CONTRIBUTING.md) and run the tests: `bash scripts/run_all_tests.sh`

2. You may need to run the linters: `bash scripts/format_backend.sh` and `bash scripts/format_frontend.sh`

akx commented on gradio/cli/commands/reload.py:
```diff
@@ -96,10 +100,23 @@ def _setup_config(
+    # so that we can reload the modules after parsing out
+    # the gr.no_reload contexts
+    for s in watching_dirs:
+        temp_dir = tempfile.mkdtemp()
+        for file in Path(s).rglob("*.py"):
+            src = file
+            dst = Path(temp_dir) / file.relative_to(s)
+            dst.parent.mkdir(parents=True, exist_ok=True)
+            shutil.copy(src, dst)
```
This will lead to interesting performance characteristics if the the tempdir ends up on a different file system, I'm sure.

freddyaboulton replied:
I realized we don't actually need to do the copying

abidlabs commented:
Would be good add gr.NO_RELOAD here: https://www.gradio.app/guides/developing-faster-with-reload-mode

freddyaboulton commented:
Open for review - can wait until the release goes out!

abidlabs commented on gradio/utils.py:
```diff
@@ -144,6 +150,51 @@ def swap_blocks(self, demo: Blocks):
+            node.body = [ast.Pass()]
+
+    # convert tree to string
+    code_removed = ast.unparse(tree)
```
It looks like `ast.unparse` was added in Python 3.9: https://docs.python.org/3/library/ast.html#ast.unparse

If we absolutely need this method, we could use this third-party implementation: https://github.com/simonpercivall/astunparse

abidlabs replied:
Btw we should confirm that the reload tests are using 3.8 (otherwise, should not have passed)

freddyaboulton replied:
Only 7 months until 3.8 EOL 🤩

I think we can manually modify the source code string so will look into it

abidlabs replied:
Thanks @freddyaboulton, ping me whenever I can take another pass at the pr

freddyaboulton replied:
@abidlabs this should be good for review now

abidlabs commented on demo/calculator/run.py:
Let's revert the changes in this file since we use this demo in many guides

abidlabs commented on gradio/networking.py:
What changed in this file?

freddyaboulton replied:
I added a new env variable at the top and the whole thing got shifted by one line. There are some minor changes to the `if GRADIO_WATCH_DIRS:` codeblock of `start_server` (basically just passing this new env variable).

abidlabs commented on guides/09_other-tutorials/developing-faster-with-reload-mode.md:
```diff
@@ -101,6 +101,31 @@ Which you could run like this: `gradio run.py --name Gretel`
+## Controlling the Reload 🎛️
+
+By default, reload mode will re-run your entire script for every change you make.
+But there are some cases where this is not desireable.
```
```suggestion
But there are some cases where this is not desirable.
```

abidlabs commented on js/_website/src/routes/[[version]]/docs/no-reload/+page.server.ts:
```diff
@@ -0,0 +1,72 @@
+	let obj = {
+		name: "NO_RELOAD",
+		description:
+			"When launching Any code in a `if gr.NO_RELOAD` code-block will not be re-evaluated when the source file is reloaded. This is helpful for importing modules that do not like to be reloaded (tiktoken, numpy) as well as database connections and long running set up code.",
```
```suggestion
			"Any code in a `if gr.NO_RELOAD` code-block will not be re-evaluated when the source file is reloaded with gradio's hot reload mode. This is helpful for importing modules that do not like to be reloaded (tiktoken, numpy) as well as database connections and long running set up code.",
```

abidlabs commented on js/_website/src/routes/[[version]]/docs/no-reload/+page.server.ts:
```diff
@@ -0,0 +1,72 @@
+		name: "NO_RELOAD",
+		description:
+			"When launching Any code in a `if gr.NO_RELOAD` code-block will not be re-evaluated when the source file is reloaded. This is helpful for importing modules that do not like to be reloaded (tiktoken, numpy) as well as database connections and long running set up code.",
+		example: `if gr.NO_RELOAD:
```
I would include a more complete example of a gradio app here to help distinguish between what should be in `if gr.NO_RELOAD` and what shouldn't

abidlabs commented on gradio/state_holder.py:
```diff
@@ -19,6 +19,10 @@ def set_blocks(self, blocks: Blocks):
         self.blocks = blocks
         self.capacity = blocks.state_session_capacity
 
+    def reset(self, blocks: Blocks):
```
Just wondering why the 2nd argument is needed? Why do you need to call `self.set_blocks(blocks)` again?

freddyaboulton replied:
The ids of the second Blocks will be different from the first one

abidlabs replied:
Got it

abidlabs commented on guides/09_other-tutorials/developing-faster-with-reload-mode.md:
```diff
@@ -101,6 +101,31 @@ Which you could run like this: `gradio run.py --name Gretel`
+
+By default, reload mode will re-run your entire script for every change you make.
+But there are some cases where this is not desireable.
+For example, loading a machine learning model should probably only happen once to save time. There are also some python libraries that use C or Rust extensions that throw errors when they are reloaded, like `numpy` and `tiktoken`.
```
```suggestion
For example, loading a machine learning model should probably only happen once to save time. There are also some Python libraries that use C or Rust extensions that throw errors when they are reloaded, like `numpy` and `tiktoken`.
```

abidlabs commented:
Very cool @freddyaboulton and easy to use!

As I was testing, I noticed two things:

1. Now, when I run a file using `gradio app.py`, the entire contents of the file get executed twice initially. To repro, create a file called `app.py` with these contents:

```py
import gradio as gr
import numpy as np

print("2") 

if gr.NO_RELOAD:
    print("1")

demo = gr.Interface(
    lambda s: s, gr.Image(np.random.random((255, 255, 3))), gr.Image()
)

if __name__ == "__main__":
    demo.launch()
```

You'll see that this is printed to the terminal:
2
1
2
1

2. I had expected that `if gr.NO_RELOAD:` is present in a file that is imported from the main file, it would also work -- since not all heavy imports are necessarily included in the main file. However, that does not seem to be the case. For example, if you have a second file called `run.py` and you import it from `app.py` with:

```
import gradio as gr

if gr.NO_RELOAD:
    print("3")
```

You'll noticed that "3" will be printed to the terminal every time that the original file is changed.

freddyaboulton commented:
Thanks for the review @abidlabs ! I've addressed all of the comments you left in the PR. For point 1, that's expected. In order to reload the modules in the thread running the reloader, the modules need to be first imported. Point 2 should be fixed now!

abidlabs reviewed (approved):
Good stuff @freddyaboulton! Love the fact that we now have a test for reload mode

freddyaboulton commented:
Thanks for the review @abidlabs ! Excited to release this
